### ISSUE TITLE
Examine indexes not written to temp folder in Azure

### ISSUE THREAD

**Issue by mortenbock** [umbraco/Umbraco-CMS#15783]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

12.0.1

### Bug summary

We're deploying to a scaled out publisher/subscriber setup on Azure, and we're seeing issues with corrupted Examine indexes.

### Specifics

We're setting this value for the publisher role, which has a single instance:

`Umbraco:CMS:Examine:LuceneDirectoryFactory=SyncedTempFileSystemDirectoryFactory`

Looking at the `LuceneIndexFolder` value for the index (see screenshot), it looks like it should be in `c:\local\temp\examineindexes\6ea291505a94cafbad209a90085b57bf\externalindex`, however, when navigating to `c:\local\temp` through Kudu, there are no examine related folders there.

The indexes DO however get created in the `C:\home\site\wwwroot\umbraco\Data\TEMP\ExamineIndexes` folder, and we are seeing exceptions like this when editors are working in the backoffice:

```
invalid deletion count: 3 vs docCount=1 (resource: BufferedChecksumIndexInput(SimpleFSIndexInput(path="C:\home\site\wwwroot\umbraco\Data\TEMP\ExamineIndexes\ExternalIndex\segments_14j")))
```

### Steps to reproduce

![image](https://github.com/umbraco/Umbraco-CMS/assets/1760256/f808a62c-5499-4650-a7f1-580a9725b724)

### Expected result / actual result

Should Umbraco not be writing indexes to the `c:\local\temp` folder, and then syncing back to the other folder?

**Comment by bielu** [comment]
@mortenbock there is no possiblity to see index in temp folder through KUDU. https://github.com/projectkudu/kudu/wiki/Understanding-the-Azure-App-Service-file-system
There are ways to disable isolation, but they might or not work in azure depends on mood of azure :)

**Comment by bielu** [comment]
is this expection on subscriber or publisher? 
As it subscribers should be in temp directory configuration, when publisher should be on sync temp :)

**Comment by mortenbock** [comment]
What is the reason for using the `SyncedTempFileSystemDirectoryFactory` on the publisher instance? If the indexes from the local temp folder are synced to the wwwroot temp folder, then when are they consumed from the wwwroot temp folder?

Is it just a way to reduce startup time if the azure instance is destroyed and the service starts on a new instance?

**Comment by bielu** [comment]
@mortenbock I dont't remember but I think it is related to stability, in case if app restarts when writing etc

**Comment by nul800sebastiaan** [comment]
Just wanted to give some background info.

> I don't understand why I'm seeing exceptions for writing to the C:\home\site\wwwroot\umbraco\Data\TEMP\ExamineIndexes index folder though.

`Synced` is in the name, and is defined as:
> The index will operate on a local index created in the processes `%temp%` location and will replicate back to main storage in 

And why this is needed: https://docs.umbraco.com/umbraco-cms/fundamentals/setup/server-setup/load-balancing/file-system-replication#examine-directory-factory-options

> This setting is needed because Lucene has issues when working from a remote file share so the files need to be read/accessed locally. Any time the index is updated, this setting will ensure that both the locally created indexes and the normal indexes are written to. This will ensure that when the app is restarted or the local environment temp files are cleared out that the index files can be restored from the centrally stored index files.

**Comment by mortenbock** [comment]
@nul800sebastiaan thank you for the update.

This should mean, that if I'm ok with rebuilding indexes on restarts, then I could just not use the Sync version of the directory factory on my publisher instance?

We're currently using deployment slots, so in any case, what is in the main storage would be outdated when we deploy, and would need rebuilding anyway.

**Comment by bielu** [comment]
@mortenbock you shouldn't use deployment slots on publisher instance. it is basically load balancing

**Comment by mortenbock** [comment]
@bielu We're moving away from it, but it's where we are now. We stop the staging slot when not deploying, so there will only be two instances for a short time, which would also be the case when instances are moved around internally by azure.

**Comment by bielu** [comment]
@mortenbock which is enough to cause issues with examine (i sadly experience it sadly)

**Comment by kevinsteffer** [comment]
I maybe have some discoveries that could lead you in a right direction regarding not being able to see the Examine index files and folders on your frontend instances (Subscribers).

When you login to KUDU the KUDU App itself doesn't share the same filesystem og C:\local\temp as the Web App does. 
That's why you can login to KUDU and see any ExamineIndexes at the location of the %TEMP% env. variable which right now is C:\local\temp. I wrote an ApiController endpoint that would send me the location of %TEMP% on the frontend instances and list it's content - and through the ApiController endpoint it returns C:\local\Temp\ExamineIndexes and all the index folders and files.

Reference: https://learn.microsoft.com/en-us/azure/app-service/operating-system-functionality#file-access-across-multiple-instances and you can read more about Files Systems in Web Apps here https://github.com/projectkudu/kudu/wiki/Understanding-the-Azure-App-Service-file-system

As part of ExamineIndexes in a load balance setup on the Subscribers and single instance on the BackOffice app I managed to have it working fine with SyncedTempFileSystemDirectoryFactory on the BackOffice App and TempFileSystemDirectoryFactory on the frontend App (Subscribers).
I run with Explicit Server Roles through some custom code that is controlled by and Appsetting configuration.

But since I've disabled the BackOffice and its endpoint on the Frontend App - I had no visual of the indexes being rebuild and when. But that's what my ApiController endpoint can tell me.

**Comment by p-m-j** [comment]
> I think I'd be afraid that some of my instances would be marked as the subscriber and not generate indexes at all.

@nul800sebastiaan if the azure load balancing guidance is being followed for frontend servers (i.e. use `TempFileSystemDirectoryFactory`) each instance will always have its own copy of the lucene index and those indexes will live at **%TEMP%/ExamineIndexes/{Some Unique Id}** which is not shared between instances.

This setup is required as the default path **/home/site/wwwroot/umbraco/Date/TEMP/ExamineIndexes** IS shared between all instances of an app service plan and lucene locking doesn't play nice on that shared file system.

On the admin server which according to the guidance should be single instance only it is safe to sync from /home/site/wwwroot as no other process should be able to write there.

> This should mean, that if I'm ok with rebuilding indexes on restarts, then I could just not use the Sync version of the directory factory on my publisher instance?

@mortenbock Yes it's a performance optimization to load the synced copy rather than rebuilding the index from scratch, if it's OK to rebuild for frontend instances it is OK for the backoffice instance also.

> you shouldn't use deployment slots on publisher instance. it is basically load balancing

@bielu In theory this should be completely fine to do as the new slot should wait to aquire main dom status from the old slot before any lucene related stuff takes place.

It's probably pragmatic advice if you have a standalone admin server that is only used for backoffice traffic, but it's perfectly reasonable to send public/anon traffic to the admin server as well as any aditional subscriber servers in which case if you still want zero downtime deployments you have to do something and a deployment slot sounds like a reasonable approach.

**Comment by bielu** [comment]
> @bielu In theory this should be completely fine to do as the new slot should wait to aquire main dom status from the old slot before any lucene related stuff takes place.
It's probably pragmatic advice if you have a standalone admin server that is only used for backoffice traffic, but it's perfectly reasonable to send public/anon traffic to the admin server as well as any aditional subscriber servers in which case if you still want zero downtime deployments you have to do something and a deployment slot sounds like a reasonable approach.

my experience telling opposite :) but yeah in theory it would be fine, but this fix it:
https://github.com/umbraco/Umbraco-CMS/pull/15571
So in theory it should be all working now ;p

**Comment by mortenbock** [comment]
>  If you think that there isn't any overhead, think about this: If you scale out to +5 nodes in a load balancing setup, that means that 5x nodes will be performing index rebuilds around the same time, this means that your DB is going to get pummeled by queries to build all of those new indexes.

The official guidance is to use the TempEnv file system for replicas, so the scale out issue will be there whether or not we use the Synced provider on the primary server.

PS: The youtube link sent me to an ad, but I did manage to find the talk :D

**Comment by Shazwazza** [comment]
Yep that's absolutely right, scaling out will always have that issue, or if your replicas are ever moved between workers. I'll check the link (dang youtube!)

**Comment by nielslyngsoe** [comment]
This issue hasn't had any activity for a long time.

Is it still relevant for you? Let us know.
